### PR TITLE
add BNL newspapers 

### PR DESCRIPTION
**Pull request by davanstrien**
This pull request adds the BNL's [processed newspaper collections](https://data.bnl.lu/data/historical-newspapers/) as a dataset. This is partly done to support BigScience see: https://github.com/bigscience-workshop/data_tooling/issues/192. 

The Datacard is more sparse than I would like but I plan to make a separate pull request to try and make this more complete at a later date. 

I had to manually add the `dummy_data` but I believe I've done this correctly (the tests pass locally).

### CODE REVIEW

**Review by albertvillanova — COMMENTED** [review]
Hi @davanstrien.

First of all, thanks a lot for your contribution. 

Within the BigScience Catalogue Data hackathon, the idea was to:
- collect all primary raw datasets (overall, for those complicated to be collected/downloaded)
- and then implement processing/parsing/conversion scripts for some of them (those requiring further processing), besides already processed datasets

So, thank you for advancing on the latter.

My initial idea was to contribute the script also as "community" datasets (instead of canonical), i.e. in this case, pushing the script to the repo https://huggingface.co/datasets/bigscience-catalogue-data/bnl_historical_newspapers 
- One of the advantages is that no dummy data is required, so the addition can be made faster
- On the other hand, one disadvantage is that contributions cannot be made through PRs
  - Therefore, we should use the Issue page for discussions, reviews, decisions,...

Maybe worth having @lhoestq 's opinion...

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (albertvillanova) [inline datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:18]: `cElementTree` is deprecated since version 3.3. Better using `ElementTree` instead?

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (albertvillanova) [inline datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:145]: No need of `os.path.join` here.

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (albertvillanova) [inline datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:158]: Better `find` instead of `findall` and then `[0]`?

Comment (albertvillanova) [inline datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:158]: Also, maybe worth passing a dict to `namespaces=` parameter instead of manually adding the explicit URI to every tag?

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (albertvillanova) [inline datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:176]: Why "ark_id" instead of "id"?

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (albertvillanova) [inline datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:174]: Maybe you missed the "language" tag? I think this information is very important, because I think they have articles in French and German (so that users could filter on this).

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Anyway, in relation with the implementation, below you can find some comments/questions.

**Review by albertvillanova — COMMENTED** [review]
Also, maybe calling the dataset as "bnl_historical_newspapers" and setting "processed" as one configuration name?

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (davanstrien) [reply datasets/bnl_processed_newspapers/bnl_processed_newspapers.py:145]: `ark_id` refers to [Archival Resource Key](https://en.wikipedia.org/wiki/Archival_Resource_Key) but probably this isn't so relevant here so I can change this to `id`

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (davanstrien) [reply datasets/bnl_processed_newspapers/bnl_processed_newspapers.py]: Yes, this is definetly an ommision. I have added this in now.

**Comment by albertvillanova** [comment]
I put here one of the source data entries, as an example:
```
<?xml version=\'1.0\' encoding=\'UTF-8\'?>
<OAI-PMH xmlns="http://www.openarchives.org/OAI/2.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/ http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd">
	<responseDate>2019-02-28T09:47:35</responseDate>
	<request>http://www.eluxemburgensia.lu/OAI</request>
	<ListRecords>
		<record>
			<header>
				<identifier>digitool-publish:1400000-DTL34</identifier>
				<datestamp>2019-02-28T09:47:35Z</datestamp>
			</header>
			<metadata>
				<oai_dc:dc xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:dcterms="http://purl.org/dc/terms/" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd">
					<dc:identifier>https://persist.lu/ark:/70795/g5rs6f/articles/DTL34</dc:identifier>
					<dc:source>newspaper/courriergdl/1844-09-04</dc:source>
					<dcterms:isPartOf>COURRIER DU GRAND-DUCHÉ DE LUXEMBOURG.</dcterms:isPartOf>
					<dcterms:isReferencedBy>issue:newspaper/courriergdl/1844-09-04/article:DTL34</dcterms:isReferencedBy>
					<dc:date>1844-09-04</dc:date>
					<dc:publisher>Imprimerie J. Lamort</dc:publisher>
					<dc:relation>1400000</dc:relation>
					<dcterms:hasVersion>http://www.eluxemburgensia.lu/webclient/DeliveryManager?pid=1400000#panel:pp|issue:1400000|article:DTL34</dcterms:hasVersion>
					<dc:description>BULLETIN POLITIQUE. Les rapports du Prince de Joinville et du maréchal Bugeun I , sur les laits d\'armes accomplis par mer cl pai terre , sont enfin publiés ; nous en donnons des extraits. Il en résulte la conséquence de la (in prochaine de la guerre du Maroc, par suite de l\'obligation où esl l\'empereur d\'accepter les conditions de la France. On ci oit assez généralement, d\'après des renseignements arrivés de divers points, que l\'empereur de Maroc a fail saisir la personne d\'Abd-rl-Kadcr, et l\'a livré au maréchal Bugeaud. Ce résultat, s\'il se confirme, contribuera , avec la victoire d\'Isly, à consolider le pouvoir de la Fiance eu Afrique. On assure que le tluc d\'Aumule va èlrc nommé gouverneur de l\'Algéiie. Le-; allaites de Taïti avancent également. S\'il laut en croire le Constitutionnel qui se prél nd bien informé, le ministèie blâmerait publiquement M. d\'Aubigny pour ses procédé» envers M. Pritchaid- il declatera que M. d\'Auliigny a clé dans son droit ni étant a M. Pritchard le pouvoir de conspirer, mais qu\'il l\'a fait avec des formes inaccoutumées, el dont la violence ne peut être approuvée. Voilà à quelle décision le cabinet se serait attelé après de longues discussions en présence de tous les ministres rappelés pour y prendre pari. M. Guizot, pour obtenir que l\'Angleterre se contente de cette déclaration , compte beaucoup sur l\'appui de lord Cowley. On assure que l\'Angleterre est disposée à reconnaître définitivement la conquête française cn Algérie, moyennant la reconnaissance de la France du protectorat augl lis sur l\'Egypte , du consentement de Méhcmtl-Ali. Le ministère fiançais a envoyé M. de Glucksbcrg à Cadix ", de là il se rendra è Tanger, comme négociateur chargé spécialement de traiter de la paix avec l\'crnpeieur de Maroc. Le voyage de Louis-Philippe, selon le Morning Post, dépendra de la tournure que prendront les négociations actuelles ; si le différend de Taïti vient tant soit peu à s\'arranger, le départ du royal visiteur sera certain.</dc:description>
					<dc:title>BULLETIN POLITIQUE.</dc:title>
					<dc:type>ARTICLE</dc:type>
					<dc:language>fr</dc:language>
					<dcterms:extent>327</dcterms:extent>
				</oai_dc:dc>
			</metadata>
		</record>
	</ListRecords>
</OAI-PMH>
```

**Review by albertvillanova — COMMENTED** [review]
I put here one of the source data entries, as an example:
```
<?xml version=\'1.0\' encoding=\'UTF-8\'?>
<OAI-PMH xmlns="http://www.openarchives.org/OAI/2.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/ http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd">
	<responseDate>2019-02-28T09:47:35</responseDate>
	<request>http://www.eluxemburgensia.lu/OAI</request>
	<ListRecords>
		<record>
			<header>
				<identifier>digitool-publish:1400000-DTL34</identifier>
				<datestamp>2019-02-28T09:47:35Z</datestamp>
			</header>
			<metadata>
				<oai_dc:dc xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:dcterms="http://purl.org/dc/terms/" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd">
					<dc:identifier>https://persist.lu/ark:/70795/g5rs6f/articles/DTL34</dc:identifier>
					<dc:source>newspaper/courriergdl/1844-09-04</dc:source>
					<dcterms:isPartOf>COURRIER DU GRAND-DUCHÉ DE LUXEMBOURG.</dcterms:isPartOf>
					<dcterms:isReferencedBy>issue:newspaper/courriergdl/1844-09-04/article:DTL34</dcterms:isReferencedBy>
					<dc:date>1844-09-04</dc:date>
					<dc:publisher>Imprimerie J. Lamort</dc:publisher>
					<dc:relation>1400000</dc:relation>
					<dcterms:hasVersion>http://www.eluxemburgensia.lu/webclient/DeliveryManager?pid=1400000#panel:pp|issue:1400000|article:DTL34</dcterms:hasVersion>
					<dc:description>BULLETIN POLITIQUE. Les rapports du Prince de Joinville et du maréchal Bugeun I , sur les laits d\'armes accomplis par mer cl pai terre , sont enfin publiés ; nous en donnons des extraits. Il en résulte la conséquence de la (in prochaine de la guerre du Maroc, par suite de l\'obligation où esl l\'empereur d\'accepter les conditions de la France. On ci oit assez généralement, d\'après des renseignements arrivés de divers points, que l\'empereur de Maroc a fail saisir la personne d\'Abd-rl-Kadcr, et l\'a livré au maréchal Bugeaud. Ce résultat, s\'il se confirme, contribuera , avec la victoire d\'Isly, à consolider le pouvoir de la Fiance eu Afrique. On assure que le tluc d\'Aumule va èlrc nommé gouverneur de l\'Algéiie. Le-; allaites de Taïti avancent également. S\'il laut en croire le Constitutionnel qui se prél nd bien informé, le ministèie blâmerait publiquement M. d\'Aubigny pour ses procédé» envers M. Pritchaid- il declatera que M. d\'Auliigny a clé dans son droit ni étant a M. Pritchard le pouvoir de conspirer, mais qu\'il l\'a fait avec des formes inaccoutumées, el dont la violence ne peut être approuvée. Voilà à quelle décision le cabinet se serait attelé après de longues discussions en présence de tous les ministres rappelés pour y prendre pari. M. Guizot, pour obtenir que l\'Angleterre se contente de cette déclaration , compte beaucoup sur l\'appui de lord Cowley. On assure que l\'Angleterre est disposée à reconnaître définitivement la conquête française cn Algérie, moyennant la reconnaissance de la France du protectorat augl lis sur l\'Egypte , du consentement de Méhcmtl-Ali. Le ministère fiançais a envoyé M. de Glucksbcrg à Cadix ", de là il se rendra è Tanger, comme négociateur chargé spécialement de traiter de la paix avec l\'crnpeieur de Maroc. Le voyage de Louis-Philippe, selon le Morning Post, dépendra de la tournure que prendront les négociations actuelles ; si le différend de Taïti vient tant soit peu à s\'arranger, le départ du royal visiteur sera certain.</dc:description>
					<dc:title>BULLETIN POLITIQUE.</dc:title>
					<dc:type>ARTICLE</dc:type>
					<dc:language>fr</dc:language>
					<dcterms:extent>327</dcterms:extent>
				</oai_dc:dc>
			</metadata>
		</record>
	</ListRecords>
</OAI-PMH>
```

**Comment by davanstrien** [comment]
> Also, maybe calling the dataset as "bnl_historical_newspapers" and setting "processed" as one configuration name?

This sounds like a good idea but my only question around this is how easy it would be to use the same approach for processing the other newspaper collections [https://data.bnl.lu/data/historical-newspapers/](). 

For example, the "BIG DATA PACK" is `257GB` of ALTO XML. This format is slightly more annoying to process because the metadata and text are contained in different files but the bigger issue might be that processing this XML using the Python XML libraries will probably be quite slow? I had thought for those larger datasets it might be more appropriate to use the Beam datasets? I don't have any experience using Beam so I'm not sure what that would involve and there is a reason to not include it in a dataset script alongside a non Beam dataset? 

If there isn't an issue with potentially later adding other datasets (which may require Beam) into the same script I'll add one config for the processed version now which leaves open the option for later adding the other datasets. If this makes sense I'll also change the name as you suggest. 

There is another dataset that could be a good candidate for inclusion here is the "Monograph Text pack" which is also processed into a simpler XML format however as the name suggests this isn't newspapers so might be confusing to include under a 'newspapers' script. One option would be to put everything under a `BNL` collection but it might be better to keep the monographs separate if they are added as a dataset so a single script doesn't end up including too much variety of content types?

**Comment by davanstrien** [comment]
> My initial idea was to contribute the script also as "community" datasets (instead of canonical), i.e. in this case, pushing the script to the repo [huggingface.co/datasets/bigscience-catalogue-data/bnl_historical_newspapers](https://huggingface.co/datasets/bigscience-catalogue-data/bnl_historical_newspapers)

Sorry to respond to this late - happy for this to go in the community datasets. I think it would be nice to include in the canonical datasets at some point but since there is less urgency with this I could try and first work on improving the Datacard before doing that (i.e. make this a draft PR) - let me know if you think that makes more sense?

**Comment by lhoestq** [comment]
> My initial idea was to contribute the script also as "community" datasets (instead of canonical), i.e. in this case, pushing the script to the repo https://huggingface.co/datasets/bigscience-catalogue-data/bnl_historical_newspapers
> One of the advantages is that no dummy data is required, so the addition can be made faster
> On the other hand, one disadvantage is that contributions cannot be made through PRs
> Therefore, we should use the Issue page for discussions, reviews, decisions,...

Sure we can use the issues to discuss/review community datasets. Maybe let's have an issue template for that ?
For this dataset in particular I'll let @albertvillanova decide whether it's best as community dataset or not. IMO both are fine :)

> I had thought for those larger datasets it might be more appropriate to use the Beam datasets? I don't have any experience using Beam so I'm not sure what that would involve and there is a reason to not include it in a dataset script alongside a non Beam dataset?

Beam is nice to process a dataset once and for all and store the resulting processed data on the Hugging Face Hub or elsewhere. However for big datasets it must run on a distributed processing runtime like Google DataFlow, which is often inconvenient for many users. We've been using it though for datasets like Wikipedia and sharing the processed data in a GCS bucket.

So feel free to use the tools you like to process the datasets, but in the end I think we just need to host the processed data in a convenient format on the Hugging Face Hub to share it with the community. The processing script you used can also be shared with the community for reproducibility and documentation. But maybe @albertvillanova already has something in mind

**Comment by davanstrien** [comment]
> > My initial idea was to contribute the script also as "community" datasets (instead of canonical), i.e. in this case, pushing the script to the repo [huggingface.co/datasets/bigscience-catalogue-data/bnl_historical_newspapers](https://huggingface.co/datasets/bigscience-catalogue-data/bnl_historical_newspapers)
> > One of the advantages is that no dummy data is required, so the addition can be made faster
> > On the other hand, one disadvantage is that contributions cannot be made through PRs
> > Therefore, we should use the Issue page for discussions, reviews, decisions,...
> 
> Sure we can use the issues to discuss/review community datasets. Maybe let's have an issue template for that ? For this dataset in particular I'll let @albertvillanova decide whether it's best as community dataset or not. IMO both are fine :)

Thanks, I'll hold off and let @albertvillanova decide best place for this. 

> > I had thought for those larger datasets it might be more appropriate to use the Beam datasets? I don't have any experience using Beam so I'm not sure what that would involve and there is a reason to not include it in a dataset script alongside a non Beam dataset?
> 
> Beam is nice to process a dataset once and for all and store the resulting processed data on the Hugging Face Hub or elsewhere. However for big datasets it must run on a distributed processing runtime like Google DataFlow, which is often inconvenient for many users. We've been using it though for datasets like Wikipedia and sharing the processed data in a GCS bucket.
> 
> So feel free to use the tools you like to process the datasets, but in the end I think we just need to host the processed data in a convenient format on the Hugging Face Hub to share it with the community. The processing script you used can also be shared with the community for reproducibility and documentation. But maybe @albertvillanova already has something in mind

That's useful, my own 2 cents are that it would make sense to do as @albertvillanova suggested and:-

- rename the dataset to 'bnl_newspapers' 
- make the 'processed dataset' a config 

I won't try and include all the other datasets now but this leaves open the option of adding those later. The actual ALTO processing should be okay to do but I think it makes sense to do this as a one-off process and make the plain text + some associated metadata available elsewere so the dataset script can be kept simple and the processing doesn't get done multiple times. 

@albertvillanova if that sounds okay I'll update pull request to include those changes.

**Comment by davanstrien** [comment]
@albertvillanova I've now created a config (currently with only one option) and renamed the dataset. This should keep the option to add other configs based on different bnl newspapers in the future.

File: datasets/bnl_newspapers/README.md
Comment (lhoestq) [inline datasets/bnl_newspapers/README.md:2]:
```suggestion
annotations_creators:
- no-annotation
```

File: datasets/bnl_newspapers/README.md
Comment (lhoestq) [inline datasets/bnl_newspapers/README.md:4]:
```suggestion
- found
```

File: datasets/bnl_newspapers/README.md
Comment (lhoestq) [inline datasets/bnl_newspapers/README.md:21]:
```suggestion
source_datasets:
- original
```

File: datasets/bnl_newspapers/README.md
Comment (lhoestq) [inline datasets/bnl_newspapers/README.md:20]:
```suggestion
- 100K<n<1M
```

File: datasets/bnl_newspapers/README.md
Comment (lhoestq) [inline datasets/bnl_newspapers/README.md:18]: I find it prettier named this way:
```suggestion
pretty_name: BnL processed newspapers
```

File: datasets/bnl_newspapers/bnl_newspapers.py
Comment (lhoestq) [inline datasets/bnl_newspapers/bnl_newspapers.py:158]: we need to open the file for it to work in streaming mode (ET.parse doesn't support URLs, while we extended `open` to support URLs)
```suggestion
            tree = ET.parse(open(xml, encoding="utf-8"))
```

**Review by lhoestq — APPROVED** [review]
It looks all good to me :)
I'll let @albertvillanova give his final review too

File: datasets/bnl_newspapers/README.md
Comment (davanstrien) [reply datasets/bnl_newspapers/README.md]: I've removed the 'processed' part but otherwise accepted this suggestion.

File: datasets/bnl_newspapers/README.md
Comment (mariosasko) [inline datasets/bnl_newspapers/README.md:197]:
``````suggestion
```
@misc{bnl_newspapers,
title={Historical Newspapers},
url={https://data.bnl.lu/data/historical-newspapers/},
author={ Bibliothèque nationale du Luxembourg},
```
``````

File: datasets/bnl_newspapers/bnl_newspapers.py
Comment (mariosasko) [inline datasets/bnl_newspapers/bnl_newspapers.py:129]:
```suggestion
            task_templates=[LanguageModeling(text_column="text")]
        )
```

This change requires regenerating the dataset_infos.json file.

**Review by mariosasko — APPROVED** [review]
A few nits from me. Other than that, looks great!

**Comment by davanstrien** [comment]
@mariosasko thanks for those suggestions

**Comment by lhoestq** [comment]
I just merged `master` into your branch to fix the CI :)

**Comment by lhoestq** [comment]
@albertvillanova do you have additional comments ? Otherwise I think this PR is ready to merge :)

File: datasets/bnl_newspapers/bnl_newspapers.py
Comment (albertvillanova) [reply datasets/bnl_newspapers/bnl_newspapers.py]: Hey @lhoestq, please take into account that we now support `ET.parse` directly in streaming mode, since this PR:
- #3476

This is useful for all community datasets, where we cannot impose our convention of using `xopen`...
:wink:

File: datasets/bnl_newspapers/README.md
Comment (albertvillanova) [inline datasets/bnl_newspapers/README.md:73]: I guess this section needs rephrasing...

File: datasets/bnl_newspapers/README.md
Comment (albertvillanova) [inline datasets/bnl_newspapers/README.md:59]:
```suggestion
- **Homepage:**: https://data.bnl.lu/data/historical-newspapers/ 
```

File: datasets/bnl_newspapers/README.md
Comment (albertvillanova) [inline datasets/bnl_newspapers/README.md:63]: Maybe an email address here? Either yours or opendata@bnl.etat.lu

File: datasets/bnl_processed_newspapers/bnl_processed_newspapers.py
Comment (albertvillanova) [reply datasets/bnl_processed_newspapers/bnl_processed_newspapers.py]: I didn't know about this ID. If you think it makes sense to the community, then you can keep it. As you prefer...

File: datasets/bnl_newspapers/README.md
Comment (albertvillanova) [inline datasets/bnl_newspapers/README.md:107]: This should be made consistent:
- either renaming it to "id"
- or renaming back the feature to "ark_id"

File: datasets/bnl_newspapers/README.md
Comment (albertvillanova) [inline datasets/bnl_newspapers/README.md:92]: Also here.

File: datasets/bnl_newspapers/README.md
Comment (albertvillanova) [inline datasets/bnl_newspapers/README.md:30]: Use the pretty name here:
```suggestion
# Dataset Card for BnL Newspapers 
```

**Review by albertvillanova — APPROVED** [review]
@davanstrien you did an awsome job!!! Thanks a lot!

Just some very minor comments (mainly about the README), and we merge this to master!

File: datasets/bnl_newspapers/README.md
Comment (davanstrien) [reply datasets/bnl_newspapers/README.md]: I have updated this to reflect the current configuration. I still hope to add the other collections as different configurations so I can update this further in a future pull request if that's okay?

**Comment by davanstrien** [comment]
> @davanstrien you did an awsome job!!! Thanks a lot!
> 
> Just some very minor comments (mainly about the README documentation), and we merge this to master!

Thanks! Hopefully all addressed now.  Thanks again for all the support with this pull request!

**Pull request merged**
